File: backend/frontend/src/services/adminFeatures.js
// Service pour gérer les fonctionnalités via l'API backend (panneau admin)
import { getFeaturesFromAPI, updateFeatureAPI, resetFeaturesAPI } from './adminFeaturesAPI';

// Fonction pour récupérer les fonctionnalités depuis l'API
export const getFeatures = async () => {
  try {
    const features = await getFeaturesFromAPI();
    return features;
  } catch (error) {
    console.error('Erreur lors du chargement des fonctionnalités:', error);
    throw error;
  }
};

// Fonction pour mettre à jour une fonctionnalité via l'API
export const updateFeature = async (featureKey, enabled) => {
  try {
    const updatedFeatures = await updateFeatureAPI(featureKey, enabled);
    
    // Déclencher un événement personnalisé pour notifier Herbbie
    window.dispatchEvent(new CustomEvent('featuresUpdated', { 
      detail: updatedFeatures 
    }));
    
    return updatedFeatures;
  } catch (error) {
    console.error('Erreur lors de la mise à jour de la fonctionnalité:', error);
    throw error;
  }
};

// Fonction pour vérifier si une fonctionnalité est activée
export const isFeatureEnabled = async (featureKey) => {
  try {
    const features = await getFeatures();
    return features[featureKey]?.enabled || false;
  } catch (error) {
    console.error(`Erreur lors de la vérification de la fonctionnalité ${featureKey}:`, error);
    return false;
  }
};

// Fonction pour récupérer les fonctionnalités activées
export const getEnabledFeatures = async () => {
  try {
    const features = await getFeatures();
    return Object.entries(features)
      .filter(([key, feature]) => feature.enabled)
      .reduce((enabled, [key, feature]) => {
        enabled[key] = feature;
        return enabled;
      }, {});
  } catch (error) {
    console.error('Erreur lors de la récupération des fonctionnalités activées:', error);
    return {};
  }
};

// Fonction pour réinitialiser toutes les fonctionnalités via l'API
export const resetFeatures = async () => {
  try {
    const resetFeaturesData = await resetFeaturesAPI();
    
    // Déclencher un événement personnalisé pour notifier Herbbie
    window.dispatchEvent(new CustomEvent('featuresUpdated', { 
      detail: resetFeaturesData 
    }));
    
    return resetFeaturesData;
  } catch (error) {
    console.error('Erreur lors de la réinitialisation des fonctionnalités:', error);
    throw error;
  }
};

// Fonction pour récupérer toutes les fonctionnalités
export const getAllFeatures = async () => {
  return await getFeatures();
};

// Fonction utilitaire pour vérifier si toutes les fonctionnalités requises sont activées
export const areRequiredFeaturesEnabled = async (requiredFeatures = []) => {
  try {
    const enabledFeatures = await getEnabledFeatures();
    return requiredFeatures.every(feature => enabledFeatures[feature]);
  } catch (error) {
    console.error('Erreur lors de la vérification des fonctionnalités requises:', error);
    return false;
  }
};

// Fonction pour synchroniser avec Herbbie (événement personnalisé)
export const syncWithHerbbie = async () => {
  try {
    const features = await getFeatures();
    
    // Déclencher un événement pour notifier Herbbie des changements
    window.dispatchEvent(new CustomEvent('herbbieFeaturesUpdate', { 
      detail: features 
    }));
    
    return true;
  } catch (error) {
    console.error('Erreur lors de la synchronisation avec Herbbie:', error);
    return false;
  }
};

// Export par défaut pour compatibilité
export default {
  getFeatures,
  updateFeature,
  isFeatureEnabled,
  getEnabledFeatures,
  resetFeatures,
  getAllFeatures,
  areRequiredFeaturesEnabled,
  syncWithHerbbie
};

File: backend/saas/routes/admin_features.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Dict, Any
from copy import deepcopy
import json
import os
from datetime import datetime

# Fichier de stockage local pour les configurations
CONFIG_FILE = "features_config.json"

router = APIRouter(prefix="/api", tags=["admin"])

# Modèles Pydantic
class FeatureUpdate(BaseModel):
    enabled: bool

class FeaturesResponse(BaseModel):
    features: Dict[str, Any]

# Configuration par défaut des fonctionnalités
DEFAULT_FEATURES = {
    "animation": {"enabled": True, "name": "Dessin animé", "icon": "🎬", "description": "Génération de dessins animés personnalisés avec IA"},
    "comic": {"enabled": True, "name": "Bande dessinée", "icon": "💬", "description": "Création de bandes dessinées avec bulles de dialogue"},
    "coloring": {"enabled": True, "name": "Coloriage", "icon": "🎨", "description": "Pages de coloriage à imprimer pour les enfants"},
    "histoire": {"enabled": True, "name": "Histoire", "icon": "📖", "description": "Histoires avec possibilité d'ajouter une narration audio"},
    "rhyme": {"enabled": True, "name": "Comptine", "icon": "🎵", "description": "Comptines musicales avec paroles et mélodies"}
}


def normalize_features_data(features: Dict[str, Any]) -> Dict[str, Any]:
    """Normaliser la configuration (supprimer les doublons, forcer l'alias audio → histoire)."""
    if not isinstance(features, dict):
        return deepcopy(DEFAULT_FEATURES)

    normalized = deepcopy(features)

    # Fusionner l'ancienne clé "audio" avec "histoire" pour rétrocompatibilité
    if "audio" in normalized:
        audio_feature = normalized.pop("audio") or {}
        histoire_feature = normalized.get("histoire", {})

        # Prioriser les informations existantes sur "histoire" tout en complétant avec celles de "audio"
        merged_feature = {
            "enabled": histoire_feature.get("enabled", audio_feature.get("enabled", True)),
            "name": histoire_feature.get("name", audio_feature.get("name", "Histoire")),
            "icon": histoire_feature.get("icon", audio_feature.get("icon", "📖")),
            "description": histoire_feature.get(
                "description",
                audio_feature.get("description", "Histoires avec possibilité d'ajouter une narration audio")
            ),
            "updated_at": histoire_feature.get("updated_at", audio_feature.get("updated_at"))
        }

        normalized["histoire"] = merged_feature

    # S'assurer que la clé "histoire" existe avec les valeurs par défaut minimales
    if "histoire" not in normalized:
        normalized["histoire"] = DEFAULT_FEATURES["histoire"].copy()

    # Garantir les propriétés essentielles
    histoire_feature = normalized["histoire"]
    histoire_feature.setdefault("enabled", True)
    histoire_feature.setdefault("name", "Histoire")
    histoire_feature.setdefault("icon", "📖")
    histoire_feature.setdefault("description", "Histoires avec possibilité d'ajouter une narration audio")

    # Conserver un ordre déterministe des fonctionnalités principales
    ordered_keys = ["animation", "comic", "coloring", "histoire", "rhyme"]
    ordered_features = {key: normalized[key] for key in ordered_keys if key in normalized}

    # Ajouter toute autre fonctionnalité personnalisée à la fin pour compatibilité future
    for key, value in normalized.items():
        if key not in ordered_features:
            ordered_features[key] = value

    return ordered_features

def load_features_config():
    """Charger la configuration depuis le fichier JSON"""
    try:
        if os.path.exists(CONFIG_FILE):
            with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
                raw_features = json.load(f)
                normalized_features = normalize_features_data(raw_features)

                # Si la normalisation a modifié les données, les persister immédiatement
                if normalized_features != raw_features:
                    with open(CONFIG_FILE, 'w', encoding='utf-8') as nf:
                        json.dump(normalized_features, nf, indent=2, ensure_ascii=False)

                return normalized_features
        else:
            # Créer le fichier avec les valeurs par défaut
            save_features_config(DEFAULT_FEATURES)
            return DEFAULT_FEATURES
    except Exception as e:
        print(f"Erreur lors du chargement de la configuration: {e}")
        return DEFAULT_FEATURES

def save_features_config(features):
    """Sauvegarder la configuration dans le fichier JSON"""
    try:
        normalized_features = normalize_features_data(features)

        with open(CONFIG_FILE, 'w', encoding='utf-8') as f:
            json.dump(normalized_features, f, indent=2, ensure_ascii=False)
        return True
    except Exception as e:
        print(f"Erreur lors de la sauvegarde de la configuration: {e}")
        return False

@router.get("/features")
async def get_features():
    """Récupérer toutes les fonctionnalités"""
    try:
        features = load_features_config()
        return features
    except Exception as e:
        print(f"Erreur lors de la récupération des fonctionnalités: {e}")
        return DEFAULT_FEATURES

@router.put("/features/{feature_key}")
async def update_feature(feature_key: str, feature_update: FeatureUpdate):
    """Mettre à jour une fonctionnalité spécifique"""
    try:
        # Charger la configuration actuelle
        features = load_features_config()
        
        # Vérifier que la fonctionnalité existe
        if feature_key not in features:
            raise HTTPException(status_code=404, detail="Fonctionnalité non trouvée")
        
        # Mettre à jour la fonctionnalité
        features[feature_key]["enabled"] = feature_update.enabled
        features[feature_key]["updated_at"] = datetime.now().isoformat()
        
        # Sauvegarder la configuration
        if save_features_config(features):
            return {"features": features}
        else:
            raise HTTPException(status_code=500, detail="Erreur lors de la sauvegarde")
            
    except HTTPException:
        raise
    except Exception as e:
        print(f"Erreur lors de la mise à jour de la fonctionnalité {feature_key}: {e}")
        raise HTTPException(status_code=500, detail="Erreur interne du serveur")

@router.post("/features/reset")
async def reset_features():
    """Réinitialiser toutes les fonctionnalités à leur état par défaut"""
    try:
        # Réinitialiser toutes les fonctionnalités à True
        features = load_features_config()
        for feature_key in features:
            features[feature_key]["enabled"] = True
            features[feature_key]["updated_at"] = datetime.now().isoformat()
        
        # Sauvegarder la configuration
        if save_features_config(features):
            return {"features": features}
        else:
            raise HTTPException(status_code=500, detail="Erreur lors de la sauvegarde")
            
    except Exception as e:
        print(f"Erreur lors de la réinitialisation des fonctionnalités: {e}")
        raise HTTPException(status_code=500, detail="Erreur interne du serveur")

@router.get("/features/{feature_key}")
async def get_feature(feature_key: str):
    """Récupérer une fonctionnalité spécifique"""
    try:
        features = load_features_config()
        
        if feature_key in features:
            return features[feature_key]
        else:
            raise HTTPException(status_code=404, detail="Fonctionnalité non trouvée")
            
    except HTTPException:
        raise
    except Exception as e:
        print(f"Erreur lors de la récupération de la fonctionnalité {feature_key}: {e}")
        raise HTTPException(status_code=500, detail="Erreur interne du serveur")